text/083-boost-stdlib-dev/text.md

```diff
@@ -0,0 +1,786 @@
+team from Coq on github) will become roots maintainer, they can of
+course opt out at any time.
+
+# Drawbacks
+
+The stdlib will no longer come bundled with Coq. It will however be
+just as easy to install via the usual package managers. More notably
+it will remain an integral part of the Coq Platform.
+
+# Alternatives
+
+The technical changes proposed here can stand on their own merits, and can be adopted without ratifying the perspective presented for motivating them. Further, each of the three splits discussed under [Packages](https://github.com/proux01/ceps/blob/boost-stdlib-dev/text/083-boost-stdlib-dev/text.md#packages) could be performed without the other ones, with or without the broader vision that led to them.
+
+1. The logical structure of the library could be split into packages, enforced by the build system, without user-visible structural changes or renaming of stdlib.
+2. The standandard library could be moved to its own repository without changing its user-facing relationship with Coq and non-stdlib packaes.
+3. Specific parts of the standard library could be spun out to their own repositories without changing the rest.
+
+Out of these, 1. has the advantage that it retains most of the technical refactoring done for this CEP, and achieves some of the technical goals listed here such as allowing users to `Require` only a known subset, while minimizing disruption to users in the face of multiple possible futures for the standard library more generally. In particular, keeping the boundaries between subparts of the standard library internal to Coq as opposed to exposing them to users would simplify later changing them.
+
+For spinning out parts of the standard library, one possible heuristic to use is that components with dedicated maintainers could become independent at the initiative of these maintainers. On the other hand, if a part of the standard library is to be separated out without a new maintainer stepping up, this could be handled as a deprecation and removal, not a refactoring.
+
+Other split in multiple repositories would be possible, could still
+happen in the future, maybe not needed now. The extreme opposite would
+be to lean toward a huge monorepo with everything from the platform,
+probably not very practical, and wouldn't bring much since the
+platform already ensures uniform releases bundling everything.
+To our knowledge the monorepo approach has exactly one advantage:
+it is easier to backport changes from one package to an upstream package.
+It also has one major constraint (considered an advantage by some and a drawback by others): it makes it harder to have the following dependencies:
+`coq-roots-A <- coq-extralib-B <- coq-roots-C`. 
+
+The most concrete technical advantage of moving stdlib to its own repo, the ability to subscribe to stdlib changes, could also be achieved using a coqbot feature that @-s a specific team of stdlib subscribers.
```

Comment:
Could you elaborate on which objectives this does or does not fullfil? I can see as fullfiled:

* enabling stdlib contributors to get only notifications about the library (and not the other Coq PRs)

and as not fullfilled:

* not having lot of OCaml code in the library repo
* not practically needing to compile Coq master to contribute to the library
* avoiding the heavy reviewing process of Coq
*  keeping all doors open with respect to "Enabling the development of an evolved or different official "stdlib""
* "Being sure that users are properly discouraged to rely on parts of the stdlib that are considered subpar, especially when recommended alternatives are available through the Coq Platform."
* enabling the use of Coq without the stdlib (for instance to run the test-suite)
* enabling more freedom on the release process of the library

Reply:
Thank you for this list of goals, somehow it seems much clearer to me than the version in the CEP.

> - not having lot of OCaml code in the library repo

I do not understand why you find this desirable, and I am worried about it. The stdlib plugins are one of the strongest advantages of Coq IMO, and I want to make sure we don't make improving them harder.

> - not practically needing to compile Coq master to contribute to the library

I had not realized this was a goal, I agree it would be nice to have. Yet compiling Coq is such a tiny fraction of the pain of compiling the CI and its dependencies that I am not sure that we should make any decisions based on this without a vision for CI.

> - avoiding the heavy reviewing process of Coq

I agree that the unavailability of reviewers is a problem, but I also find code review valuable. I there anything else you definitely want to see changed?

> - keeping all doors open with respect to "Enabling the development of an evolved or different official "stdlib""

I don't see how this needs any action, the doors are already open.

> - "Being sure that users are properly discouraged to rely on parts of the stdlib that are considered subpar, especially when recommended alternatives are available through the Coq Platform."

We have warnings for that, and we can and have removed content that we don't recommend. I don't think we need another solution.

> - enabling the use of Coq without the stdlib (for instance to run the test-suite)

I agree that this is desirable, but the test suite is such a small fraction of the CI that I am not sure whether this matters (and the dune-based plan seems more appealing to me as long as coq itself is built with dune).

> - enabling more freedom on the release process of the library

Again, I am worried that exercising this freedom would put stress on the plugins that I use Coq for.

Reply:
> We don't want to make it look like mastering OCaml should be anyhow needed to contribute to the library.

Do you think there is a reason why this couldn't be straightforwardly clarified through documentation?

> The future is probably more in the use of dedicated metalanguages like metacoq, Ltac2 or elpi.

Maybe, but I don't see how these new languages would help specifically the stdlib plugins. E.g. elpi has "Lambda" in the name, and lia/field/nsatz do not manipulate any lambdas. Sure, there is some maintenance for with the ML api getting new fancier universes and so on, but constructing application nodes surely isn't changing so unstable that it needs to be avoided. I also expect multiplicative costs with both number of repositories and number of languages, so I would expect we'd do a CEP and consider alternatives for adding any of these to the standard library.

> What makes plugin development hard is the unstable, more or less documented, API, much more than their interaction with the library. So keeping them in Coq repo shouldn't be much trouble.

I agree this is a problem at least for some plugins, but it is not the problem that I was trying to point out. Each plugin exists as a combination of a certificate generator written in ML (perhaps later in another language) and a certificate checker written in Gallina. The interface between them isn't designed for wider consumption, and I have found the documentation to be wrong when I have tried relying on it. Having that interface straddle repositories with separate release schedules would require essentially any improvement to it to have a deprecation delay. This seems quite important to avoid.

> Imagine you're a newcomer and want to add some list lemma, you certainly don't want to compile Coq master and you'd very probably don't have to run any of the CI targets locally since it would just be green online.

Sadly, no. It is especially the simple stuff for which a large part of the effort is multirepo coordination overhead. Often it's just because morally equivalent lemmas with the same name already exist downstream, differing perhaps in names of binders only. [nth_error](https://github.com/coq/coq/pull/18563) was basically as easy as it gets, and I don't want to imagine doing it with web-interface-driven CI.

> No, imagine we choose to go for the "different official "stdlib"" path (I'm not saying we should, I don't know), then we would still have to move the current library to keep maintaining it, so we would be back to the beginning, with that door remaining closed.

Even a well-resourced effort for a total makeover could happen in the same repo. I think what you are saying is that if we want to deprecate the stdlib without porting users, moving it to a separate repo would be the way to do it. Perhaps that's right, but I don't want us to take that path, and making way for it isn't compelling as motivation for this CEP.

> I tend to think that displaying packages, to a proper library, on an equal footing with the alternatives, would help, compared to one of the alternatives being forced the default.

Yeah, we disagree here. IMO it is on us as stdlib maintainers to pick a default and own it. I am especially wary of throwing out old theory/plugins/languages without a replacement that is proven on the same tasks.

> At the very worst, we could imagine schemes were the plugin would have to run alternative code for two different versions of the library, but that need will likely arise very rarely.

How do you imagine this would work?

Reply:
> plugins like paramcoq tend to get quite a lot of overlays

Paramcoq seems vastly more complex in its use of the Coq API than nsatz and micromega to me.

> making a few changes to the micromega plugin in order to make it easily and effectively reusable in a foreign library proved pretty easy and stable enough to not require any change over many Coq versions. So I'm not really worried here, quite the contrary.

My different experience is also with micromega: https://github.com/coq/coq/issues/16862, https://github.com/coq/coq/pull/18730#issuecomment-1978358197.

> [on deprecating stdlib] I understand you are strongly against this possibility, but you cannot deny that not enabling it is keeping doors closed.

But I will deny it: nothing prevents one from making a PR to deprecate any fraction of the stdlib where it sits today. Or is the difference about who would review that PR and how we would expect them to respond?

> "if (new_gallina_constant is present) then new_feature"

Thanks, I agree this could work.

Reply:
> [on deprecating stdlib]

I fail to understand why you keep bringing "deprecating stdlib" whereas there is no such thing in the CEP.